### ISSUE TITLE
daemon processes need to continually attempt to restart

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2415]
_Imported from trac ticket 2415.
Created by **kbower**
Opened in trac: Jul 16, 2013, 12:26:00 PM
Last modified in trac: Jul 19, 2013, 7:59:05 AM_
***
These processes die and if supervisord can't restart within several attempts, it stops trying.

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 16, 2013, 12:27:13 PM:_

Rob, my instinct here is that this needs to be supervisor that tries to restart these, not the process itself.

**Comment by kentbower** [comment]
_**rneville** trac comment on Jul 16, 2013, 12:54:46 PM:_

Agreed. Perhaps that should be the default behavior and then a flag set in app.ini to indicate instances where you do not want that behavior (can't think of any but may want it).

**Comment by kentbower** [comment]
_**kbower** trac comment on Jul 19, 2013, 7:59:05 AM:_

(Changeset [[4358]](../commit/98c66f2989e34cc9ced618d4a86183beb02682dc))
* fixes #2415, daemon processes need to continually attempt to restart #all

 * upgrade and modify supervisor to honor expected exitcodes as clean exit even if exit is before startsecs
   * see https://github.com/Supervisor/supervisor/issues/260

```diff
diff -U10 -r supervisor-3.0b2/supervisor/process.py supervisor-3.0b201kb/supervisor/process.py
--- supervisor-3.0b2/supervisor/process.py  2013-05-28 13:15:24.000000000 -0400
+++ supervisor-3.0b201kb/supervisor/process.py  2013-07-17 17:01:39.000000000 -0400
@@ -439,21 +439,22 @@
             # likely the result of a stop request
             # implies STOPPING -> STOPPED
             self.killing = 0
             self.delay = 0
             self.exitstatus = es
 
             msg = "stopped: %s (%s)" % (processname, msg)
             self._assertInState(ProcessStates.STOPPING)
             self.change_state(ProcessStates.STOPPED)
 
-        elif tooquickly:
+        # kb added "and not exit_expected"
+        elif tooquickly and not exit_expected:
             # the program did not stay up long enough to make it to RUNNING
             # implies STARTING -> BACKOFF
             self.exitstatus = None
             self.spawnerr = 'Exited too quickly (process log may have details)'
             msg = "exited: %s (%s)" % (processname, msg + "; not expected")
             self._assertInState(ProcessStates.STARTING)
             self.change_state(ProcessStates.BACKOFF)
 
         else:
             # this finish was not the result of a stop request, the
diff -U10 -r supervisor-3.0b2/supervisor/rpcinterface.py supervisor-3.0b201kb/supervisor/rpcinterface.py
--- supervisor-3.0b2/supervisor/rpcinterface.py 2013-05-26 22:39:09.000000000 -0400
+++ supervisor-3.0b201kb/supervisor/rpcinterface.py 2013-07-17 22:21:13.000000000 -0400
@@ -291,20 +291,24 @@
                 started.append(time.time())
 
             if not wait or not startsecs:
                 return True
 
             t = time.time()
             runtime = (t - started[0])
             state = process.get_state()
 
             if state not in (ProcessStates.STARTING, ProcessStates.RUNNING):
+                # start kb, expected exit before startsecs
+                if state == ProcessStates.EXITED and process.spawnerr is None:
+                    return True
+                # stop kb
                 raise RPCError(Faults.ABNORMAL_TERMINATION, name)
 
             if runtime < startsecs:
                 return NOT_DONE_YET
 
             if state == ProcessStates.RUNNING:
                 return True
 
             raise RPCError(Faults.ABNORMAL_TERMINATION, name)
 
diff -U10 -r supervisor-3.0b2/supervisor/version.txt supervisor-3.0b201kb/supervisor/version.txt
--- supervisor-3.0b2/supervisor/version.txt 2013-05-29 01:10:37.000000000 -0400
+++ supervisor-3.0b201kb/supervisor/version.txt 2013-07-17 17:00:27.000000000 -0400
@@ -1 +1 @@
-3.0b2
+3.0b201kb
```

**Comment by kentbower** [comment]
_**kbower** trac status change on Jul 19, 2013, 7:59:05 AM:_ **closed**
